pkg/reconciler/ingress/ingress.go

```diff
@@ -249,38 +247,37 @@ func (r *Reconciler) reconcileCertSecrets(ctx context.Context, ia *v1alpha1.Ingr
 }
 
 func (r *Reconciler) reconcileVirtualServices(ctx context.Context, ia *v1alpha1.Ingress,
-	desired []*v1alpha3.VirtualService) error {
+	desired map[string]*v1alpha3.VirtualService) error {
+	// Sort VirtualService names for a consistent processing order.
+	names := sets.NewString()
+	for name := range desired {
+		names.Insert(name)
+	}
 	// First, create all needed VirtualServices.
-	kept := sets.NewString()
-	for _, d := range desired {
-		if _, err := istioaccessor.ReconcileVirtualService(ctx, ia, d, r); err != nil {
+	for _, name := range names.List() {
+		vs := desired[name]
+		if vs == nil {
+			continue
+		}
+		if _, err := istioaccessor.ReconcileVirtualService(ctx, ia, vs, r); err != nil {
 			if kaccessor.IsNotOwned(err) {
-				ia.Status.MarkResourceNotOwned("VirtualService", d.Name)
+				ia.Status.MarkResourceNotOwned("VirtualService", vs.Name)
 			}
 			return err
 		}
-		kept.Insert(d.Name)
 	}
 
 	// Now, remove the extra ones.
-	vses, err := r.virtualServiceLister.VirtualServices(resources.VirtualServiceNamespace(ia)).List(
-		labels.Set(map[string]string{
-			serving.RouteLabelKey:          ia.GetLabels()[serving.RouteLabelKey],
-			serving.RouteNamespaceLabelKey: ia.GetLabels()[serving.RouteNamespaceLabelKey]}).AsSelector())
-	if err != nil {
-		return fmt.Errorf("failed to get VirtualServices: %w", err)
-	}
-	for _, vs := range vses {
-		n, ns := vs.Name, vs.Namespace
-		if kept.Has(n) {
-			continue
-		}
-		if !metav1.IsControlledBy(vs, ia) {
-			// We shouldn't remove resources not controlled by us.
+	for _, name := range names.List() {
+		vs := desired[name]
+		if vs != nil {
 			continue
 		}
-		if err = r.IstioClientSet.NetworkingV1alpha3().VirtualServices(ns).Delete(n, &metav1.DeleteOptions{}); err != nil {
-			return fmt.Errorf("failed to delete VirtualService: %w", err)
+		ns := ia.Namespace
+		if err := r.IstioClientSet.NetworkingV1alpha3().VirtualServices(ns).Delete(name, &metav1.DeleteOptions{}); err != nil {
```

Comment:
Why did we remove the ownership check?

Reply:
We needed it previously because we perform a list.  We have changed to always return the full map of what to be created and what to be deleted (since we know the exact names of VirtualServices owned by a particular Ingress).